### ISSUE TITLE
Memory warning and crash when launch cameraView many times.

### ISSUE THREAD

**Issue by MichaelCosmo** [AlexLittlejohn/ALCameraViewController#89]
Hi, 
I have an issue with AlCameraViewController. I use this code to launch my cameraView :
`let cameraViewController = CameraViewController(croppingEnabled: croppingEnabled, allowsLibraryAccess: libraryEnabled) { image, asset in
         ...}`

When I try to launch and close the cameraView 3 times, my app crash. I have memory warnings in my Xcode console. 

There is a way to launch as many times I want my cameraView without crash my app? 

Thanks :)

**Comment by matart15** [comment]
Same here.
I think there is Strong Reference Cycle for Closures in imagePickerViewController.
After adding `[weak imagePicker]`  in these line deinit is called properly.
https://github.com/AlexLittlejohn/ALCameraViewController/blob/master/ALCameraViewController/ViewController/CameraViewController.swift#L24
https://github.com/AlexLittlejohn/ALCameraViewController/blob/master/ALCameraViewController/ViewController/CameraViewController.swift#L27

I hope it gets fixed soon

**Comment by MichaelCosmo** [comment]
How do you exactly fixed it, doesn't work with me :(

**Comment by matart15** [comment]
It is CameraViewController.imagePickerViewController function.
Inside it `imagePicker.onSelectionComplete`  and `confirmController.onComplete` closures are referencing `imagePicker`. So it never gets cleaned from memory
You can read about it in here https://developer.apple.com/library/ios/documentation/Swift/Conceptual/Swift_Programming_Language/AutomaticReferenceCounting.html

Anyway I created pull request.

**Comment by MichaelCosmo** [comment]
Ok thank you for details :)

**Comment by AlexLittlejohn** [comment]
Can you supply more code to show exactly what you are doing once you've closed the camera?

**Comment by MichaelCosmo** [comment]
Here I launch the camera => https://github.com/Yagora/CEOS_IOS/blob/master/CEOS/MapViewController.swift#L180
And, when the cameraView is dismissed I launch a popUp view here (only if a photo was chosen) => https://github.com/Yagora/CEOS_IOS/blob/master/CEOS/MapViewController.swift#L150

**Comment by AlexLittlejohn** [comment]
Ok, so a few things. 

1. It's not necessary to retain the camera view controller yourself when presenting it.
2. use `[weak self]` inside the camera completion block

e.g. 
```swift
let cameraViewController = CameraViewController(croppingEnabled: croppingEnabled) { [weak self] image, asset in
    if ((image) != nil) {
        self?.imageLoki = image
        self?.popUpShow = true
    } else {
        self?.cameraShow = false
    }
    self?.dismissViewControllerAnimated(true, completion: nil)
}

presentViewController(cameraViewController, animated: true, completion: nil)
```

These things together would have most likely caused each previous camera instance to remain in memory.

**Comment by MichaelCosmo** [comment]
Ok thank you, I'll try it soon :)

**Comment by MichaelCosmo** [comment]
It seems to didn't resolve my issue. If I launch 3 times the CameraView, my app crash and I have this message error on Xcode : [http://i.stack.imgur.com/C1Jgt.png](url)

Any idea? :(

**Comment by AlexLittlejohn** [comment]
The link to the image is broken?

**Comment by MichaelCosmo** [comment]
[And here ?](http://i.stack.imgur.com/C1Jgt.png)

**Comment by AlexLittlejohn** [comment]
Does the console print out a memory warning? Have you measured the memory usage in instruments? I ask because this doesn't look like a typical memory error.

**Comment by MichaelCosmo** [comment]
My console print out a memory warning after I've close the cameraView  and I've not used memory instrument, but I will ;)

**Comment by MichaelCosmo** [comment]
Which kind on instrument I should use (I already tested Leaks, Activity Monitor and System Usage) ? What can I do to get info on that crash?

**Comment by matart15** [comment]
for me Allocations from instrument shows me ALCameraViewController instance after i choose picture or closed ALCameraViewController
this method displays message when object is being destroyed.
`deinit {
    debugPrint("deinit: ClassName")
}`

Then I used old way to find problem.
I commented out all code inside functions until message displayed.
then put codes back to identify the exact place.

little hard way
